Index: app/src/main/java/com/jikexueyuan/bluetoothchat_1051/BluetoothConnection.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/com/jikexueyuan/bluetoothchat_1051/BluetoothConnection.java	(revision )
+++ app/src/main/java/com/jikexueyuan/bluetoothchat_1051/BluetoothConnection.java	(revision )
@@ -1,0 +1,0 @@
